File: AlgoAnalysis-C/HeapSort/heapSort.c
/*
Yapılacaklar:
1. Yığın Sıralaması (Heap Sort) için gerekli olan 'Heap' struct yapısı tanımlanacak.
2. 'Max_Heapify' fonksiyonu yazılacak: Bu fonksiyon, bir alt ağacın
   Maks-Yığın özelliğini korumasını sağlar.
3. 'Build_Max_Heap' fonksiyonu yazılacak: Bu fonksiyon, rastgele bir diziyi
   'Max_Heapify' kullanarak bir Maks-Yığına dönüştürür.
4. 'Heap_Sort' fonksiyonu yazılacak: Bu fonksiyon, 'Build_Max_Heap'
   ile yığını kurar ve ardından diziyi yerinde (in-place) sıralar.
5. main() fonksiyonuna Heap_Sort zamanlaması eklenecek.
*/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <limits.h>

#define ARRAYSIZE 100000

// Yığın (Heap) veri yapısını tanımlıyoruz.
// 'A' dizinin kendisini (veriyi) işaret eder.
// 'size' yığının o anki mantıksal boyutunu tutar.
typedef struct Heap
{
    int size;
    int *A;
}Heap;

// Fonksiyon prototipleri
void InsertionSort(int A[]);
void Merge(int A[], int p, int q, int r);
void MergeSort(int A[],int p, int r);

void QUICKSORT(int A[], int p, int r);
int PARTITION(int A[], int p, int r);

void RANDOMIZED_QUICKSORT(int A[], int p, int r);
int RANDOMIZED_PARTITION(int A[], int p, int r);

// --- Yığın Sıralaması Protototipleri ---
void Max_Heapify(Heap myHeap, int i);
void Build_Max_Heap(Heap myHeap);
void Heap_Sort(Heap myHeap);

int main()
{
    // 4 algoritma için 4 kopya dizi
    int A1[ARRAYSIZE];
    int A2[ARRAYSIZE];
    int A3[ARRAYSIZE];
    int A4[ARRAYSIZE];
    int i;
    int num;
    time_t t;
    srand((unsigned) time(&t)); // Rastgele sayı üretecini başlat

    // Tüm dizileri aynı rastgele sayılarla doldur
    for(i = 0; i < ARRAYSIZE; i++)
    {
        num = rand();
        A1[i] = num;
        A2[i] = num;
        A3[i] = num;
        A4[i] = num;
    }

    clock_t start_t, end_t;
    double cpu_time_used;

    // 1. Eklemeli Sıralama Testi
    start_t = clock();
    InsertionSort(A1);
    end_t = clock();
    cpu_time_used = (double)(end_t - start_t) / CLOCKS_PER_SEC;
    printf("Toplam gecen sure (Eklemeli Siralama): %f saniyedir.\n", cpu_time_used);

    // 2. Birleştirmeli Sıralama Testi
    start_t = clock();
    MergeSort(A2,0,ARRAYSIZE-1);
    end_t = clock();
    cpu_time_used = (double)(end_t - start_t) / CLOCKS_PER_SEC;
    printf("Toplam gecen sure (Birlestirmeli Siralama): %f saniyedir.\n", cpu_time_used);

    // 3. Hızlı Sıralama Testi
    start_t = clock();
    QUICKSORT(A3,0,ARRAYSIZE-1);
    end_t = clock();
    cpu_time_used = (double)(end_t - start_t) / CLOCKS_PER_SEC;
    printf("Toplam gecen sure (Hizli Siralama): %f saniyedir.\n", cpu_time_used);

    // 4. Yığın Sıralaması Testi
    start_t = clock();
    Heap myHeap;      // Yığın yapısını oluştur
    myHeap.A = A4;     // Yığının dizisi A4'ü göstersin
    myHeap.size = ARRAYSIZE; // Yığının başlangıç boyutu
    Heap_Sort(myHeap); // Sıralamayı başlat
    end_t = clock();
    cpu_time_used = (double)(end_t - start_t) / CLOCKS_PER_SEC;
    printf("Toplam gecen sure (Yigin Siralama): %f saniyedir.\n", cpu_time_used);

    return 0;
}

// ... InsertionSort, MergeSort, Merge, QUICKSORT, PARTITION,
// ... RANDOMIZED_QUICKSORT, RANDOMIZED_PARTITION fonksiyonları
// ... (Bu fonksiyonlar önceki adımlarda mevcuttu, değişiklik yok)

void InsertionSort(int A[])
{
    int i,j,key;
    for(j = 1; j < ARRAYSIZE; j++)
    {
        key = A[j];
        i = j - 1;
        while(i >= 0 && A[i] > key)
        {
            A[i+1] = A[i];
            i = i - 1;
        }
        A[i+1] = key;
    }
}

void Merge(int A[], int p, int q, int r)
{
    int i,j,k;
    int n1 = q - p + 1;
    int n2 = r - q;

    int *L = (int *)malloc((n1+1)*sizeof(int));
    int *R = (int *)malloc((n2+1)*sizeof(int));

    for(i = 0; i < n1; i++)
    {
        L[i] = A[p+i];
    }
    for(j = 0; j < n2; j++)
    {
        R[j] = A[q+j+1];
    }
    L[n1] = INT_MAX;
    R[n2] = INT_MAX;
    i = 0;
    j = 0;
    for(k = p; k <= r; k++)
    {
        if(L[i] <= R[j])
        {
            A[k] = L[i];
            i++;
        }
        else
        {
            A[k] = R[j];
            j++;
        }
    }
    free(L); // Ayrılan hafızayı serbest bırak
    free(R); // Ayrılan hafızayı serbest bırak
}

void MergeSort(int A[],int p, int r)
{
    int q;
    if(p < r)
    {
        q = (p + r) / 2;
        MergeSort(A,p,q);
        MergeSort(A,q+1,r);
        Merge(A,p,q,r);
    }
}

void QUICKSORT(int A[], int p, int r)
{
    int q;
    if(p < r)
    {
       q = PARTITION(A, p, r);
       QUICKSORT(A, p, q-1);
       QUICKSORT(A, q+1, r);
    }
}

int PARTITION(int A[], int p, int r)
{
    int x = A[r];
    int i = p - 1;
    int j;
    int temp;
    for(j = p; j < r; j++)
    {
       if(A[j] <= x)
       {
          i = i + 1;
          temp = A[i];
          A[i] = A[j];
          A[j] = temp;
       }
    }
    temp = A[i+1];
    A[i+1] = A[r];
    A[r] = temp;
    return i+1;
}

void RANDOMIZED_QUICKSORT(int A[], int p, int r)
{
    int q;
    if(p < r)
    {
       q = RANDOMIZED_PARTITION(A, p, r);
       RANDOMIZED_QUICKSORT(A, p, q-1);
       RANDOMIZED_QUICKSORT(A, q+1, r);
    }
}

int RANDOMIZED_PARTITION(int A[], int p, int r)
{
    int i = (rand() % (r - p + 1)) + p;
    int temp = A[r];
    A[r] = A[i];
    A[i] = temp;
    return PARTITION(A,p,r);
}

// --- Yığın Sıralaması Fonksiyonları (Yorumlu) ---

/**
 * @brief Bir alt ağacın 'i' indeksindeki kök düğümünde
 * Maks-Yığın (Max-Heap) özelliğini korur.
 * Varsayım: 'i'nin sol ve sağ alt ağaçları zaten birer Maks-Yığındır.
 * * @param myHeap Yığın yapısı (diziyi ve boyutunu içerir).
 * @param i Maks-Yığın özelliğinin kontrol edileceği kök indeksi.
 */
void Max_Heapify(Heap myHeap, int i){
    // Not: C'de diziler 0 tabanlıdır.
    // Bir 'i' düğümünün sol çocuğu: 2*i + 1
    // Bir 'i' düğümünün sağ çocuğu: 2*i + 2
    int left = 2 * i + 1;
    int right = 2 * i + 2;
    int largest = i; // En büyük elemanın indeksini (başlangıçta kök) tutar.

    // 1. Sol çocuk var mı (yığın sınırları içinde mi) VE
    //    sol çocuk kökten (A[i]) büyük mü?
    if (left < myHeap.size && myHeap.A[left] > myHeap.A[i]) {
       largest = left;
    }
    else {
       largest = i;
    }

    // 2. Sağ çocuk var mı (yığın sınırları içinde mi) VE
    //    sağ çocuk o anki 'largest' (kök veya sol çocuk) elemandan büyük mü?
    if (right < myHeap.size && myHeap.A[right] > myHeap.A[largest]) {
       largest = right;
    }

    // 3. Eğer en büyük eleman kök (i) değilse (yani çocuklardan biri büyükse):
    if (largest != i) {
       // Kök ile en büyük çocuğu yer değiştir (swap).
       int temp = myHeap.A[i];
       myHeap.A[i] = myHeap.A[largest];
       myHeap.A[largest] = temp;

       // Yer değiştirme işlemi alt ağacın yığın özelliğini bozmuş olabilir.
       // Bu yüzden, yer değiştirilen çocuğun indiği yeni konum (largest)
       // için Max_Heapify'ı özyineli olarak tekrar çağır.
       Max_Heapify(myHeap, largest);
    }
}

/**
 * @brief Rastgele dolu bir diziyi bir Maks-Yığın haline getirir.
 * Strateji: Dizideki son 'parent' (ebeveyn) düğümden başlayarak
 * geriye doğru (köke kadar) tüm ebeveyn düğümler için Max_Heapify çağırılır.
 * * @param myHeap Yığın yapısı.
 */
void Build_Max_Heap(Heap myHeap){
    // n elemanlı (0 tabanlı) bir dizide, son ebeveynin indeksi
    // (myHeap.size / 2) - 1 'dir.
    // (myHeap.size / 2)'den başlamak da çalışır (tamsayı bölmesi nedeniyle).
    // Kodunuza sadık kalıyoruz:
    for (int i = myHeap.size / 2; i >= 0; i--) {
        Max_Heapify(myHeap, i);
    }
}

/**
 * @brief Yığın Sıralaması algoritmasını çalıştırır (yerinde sıralama).
 * * @param myHeap Yığın yapısı.
 */
void Heap_Sort(Heap myHeap){
    // 1. Adım: Diziyi bir Maks-Yığın'a dönüştür.
    // Bu adımdan sonra, dizinin en büyük elemanı A[0]'dadır.
    Build_Max_Heap(myHeap);

    // 2. Adım: Yığından elemanları tek tek çıkar ve sırala
    // Dizinin son elemanından (ARRAYSIZE-1) başlayarak
    // ikinci elemanına (indeks 1) kadar döngüye gir.
    for (int i = ARRAYSIZE - 1; i >= 1; i--) {

       // a. En büyük elemanı (A[0]) al, dizinin sonuna (A[i]) taşı.
       // A[i]'deki elemanı da A[0]'a taşı (bu, yığını geçici olarak bozar).
       int temp = myHeap.A[0];
       myHeap.A[0] = myHeap.A[i];
       myHeap.A[i] = temp;

       // b. Yığının mantıksal boyutunu 1 azalt.
       // (A[i] artık sıralanmış bölgededir ve yığının parçası değildir).
       myHeap.size = myHeap.size - 1;

       // c. Yeni kök (A[0]) için Max_Heapify çağırarak
       //    yığın özelliğini onar.
       Max_Heapify(myHeap, 0);
    }
}

File: AlgoAnalysis-Python/HeapSort/HeapSort.py
import random
import time
import math  # MergeSort'taki 'sonsuz' (INT_MAX) değeri için
import sys

# Python'un özyineleme (recursion) limitini artıralım
# Quicksort ve Mergesort'un 100.000 eleman için çalışması gerekir.
# C'deki 'segmentation fault' hatasının Python'daki karşılığıdır.
sys.setrecursionlimit(200000)

# C kodundaki #define ARRAYSIZE 100000
ARRAY_SIZE = 200


# --- InsertionSort Fonksiyonu ---

def InsertionSort(A):
    """C'deki InsertionSort fonksiyonunun Python karşılığı."""
    # ARRAYSIZE yerine len(A) kullanmak daha esnektir
    for j in range(1, len(A)):
        key = A[j]
        i = j - 1
        while i >= 0 and A[i] > key:
            A[i + 1] = A[i]
            i = i - 1
        A[i + 1] = key


# --- MergeSort Fonksiyonları ---

def Merge(A, p, q, r):
    """C'deki Merge fonksiyonunun Python karşılığı."""
    n1 = q - p + 1
    n2 = r - q

    # C'deki L ve R dizileri (Python'da list slicing ile)
    L = A[p: q + 1]
    R = A[q + 1: r + 1]

    # C'deki INT_MAX (nöbetçi/sentinel) değerleri
    L.append(math.inf)
    R.append(math.inf)

    i = 0
    j = 0
    # C'deki for(k = p; k <= r; k++)
    for k in range(p, r + 1):
        if L[i] <= R[j]:
            A[k] = L[i]
            i += 1
        else:
            A[k] = R[j]
            j += 1

    # Python'da C'deki gibi 'free(L)' ve 'free(R)' yapmaya gerek yoktur.


def MergeSort(A, p, r):
    """C'deki MergeSort fonksiyonunun Python karşılığı."""
    if p < r:
        q = (p + r) // 2  # Python'da tam sayı bölmesi
        MergeSort(A, p, q)
        MergeSort(A, q + 1, r)
        Merge(A, p, q, r)


# --- Quicksort Fonksiyonları ---

def QUICKSORT(A, p, r):
    """C'deki QUICKSORT fonksiyonunun Python karşılığı."""
    if p < r:
        q = PARTITION(A, p, r)
        QUICKSORT(A, p, q - 1)
        QUICKSORT(A, q + 1, r)


def PARTITION(A, p, r):
    """C'deki PARTITION fonksiyonunun Python karşılığı."""
    x = A[r]
    i = p - 1
    # C'deki for(j = p; j < r; j++)
    for j in range(p, r):
        if A[j] <= x:
            i = i + 1
            # Python'da temp kullanmadan yer değiştirme (swap)
            A[i], A[j] = A[j], A[i]

    # Pivotu (x) doğru yerine taşı
    A[i + 1], A[r] = A[r], A[i + 1]
    return i + 1


# --- Randomized Quicksort Fonksiyonları ---

def RANDOMIZED_QUICKSORT(A, p, r):
    """C'deki RANDOMIZED_QUICKSORT fonksiyonunun Python karşılığı."""
    if p < r:
        q = RANDOMIZED_PARTITION(A, p, r)
        RANDOMIZED_QUICKSORT(A, p, q - 1)
        RANDOMIZED_QUICKSORT(A, q + 1, r)


def RANDOMIZED_PARTITION(A, p, r):
    """C'deki RANDOMIZED_PARTITION fonksiyonunun Python karşılığı."""
    # C'deki (rand() % (r - p + 1)) + p ifadesi
    i = random.randint(p, r)
    # Pivotu (rastgele seçileni) son elemanla yer değiştir
    A[r], A[i] = A[i], A[r]
    # Standart PARTITION'ı çağır
    return PARTITION(A, p, r)


# --- Heap Sort Fonksiyonları ---
def Max_Heapify(A, heap_size, i):
    """
    Bir alt ağacın 'i' indeksindeki kök düğümünde
    Maks-Yığın (Max-Heap) özelliğini korur (onarmak için kullanılır).

    Varsayım: 'i'nin sol ve sağ alt ağaçları zaten birer Maks-Yığındır.

    Parametreler:
    A (list): Dizimiz (yığını temsil eder).
    heap_size (int): Yığının mevcut (mantıksal) boyutu (dizinin tamamı olmayabilir).
    i (int): Maks-Yığın özelliğinin kontrol edileceği kök indeksi.
    """

    # 0-tabanlı indeksleme için sol ve sağ çocukların indekslerini hesapla
    left = 2 * i + 1
    right = 2 * i + 2

    # Başlangıçta en büyük elemanın kök (i) olduğunu varsay
    largest = i

    # 1. Sol çocuk var mı (yığın sınırları içinde mi) VE
    #    sol çocuk kökten büyük mü?
    if left < heap_size and A[left] > A[largest]:
        largest = left
    else:
        # Bu 'else' bloğu aslında gereksizdir, çünkü 'largest' zaten
        # 'i' olarak ayarlanmıştı. C kodundan kalma olabilir.
        largest = i

    # 2. Sağ çocuk var mı (yığın sınırları içinde mi) VE
    #    sağ çocuk o anki 'largest' (kök veya sol çocuk) elemandan büyük mü?
    if right < heap_size and A[right] > A[largest]:
        largest = right

    # 3. Eğer en büyük eleman kök (i) değilse (yani çocuklardan biri büyükse):
    if largest != i:
        # Kök ile en büyük çocuğu yer değiştir (swap).
        temp = A[i]
        A[i] = A[largest]
        A[largest] = temp

        # Yer değiştirme işlemi alt ağacın yığın özelliğini bozmuş olabilir.
        # Bu yüzden, yer değiştirilen çocuğun indiği yeni konum (largest)
        # için Max_Heapify'ı özyineli olarak tekrar çağır.
        Max_Heapify(A, heap_size, largest)


def Build_Max_Heap(A, heap_size):
    """
    Rastgele dolu bir diziyi (A) bir Maks-Yığın haline getirir.

    Strateji: Dizideki son 'parent' (ebeveyn) düğümden başlayarak
    geriye doğru (köke kadar) tüm ebeveyn düğümler için
    Max_Heapify çağırılır.
    """

    # 0-tabanlı bir dizide, son ebeveynin indeksi (heap_size // 2 - 1)'dir.
    # Bu döngü, son ebeveynden başlayarak kök düğüme (indeks 0) kadar
    # geriye doğru gider.
    for i in range(heap_size // 2 - 1, -1, -1):
        # Her ebeveyn düğüm için Max_Heapify çağırarak
        # o düğümün alt ağacının yığın özelliğini sağladığından emin ol.
        Max_Heapify(A, heap_size, i)


def Heap_Sort(A):
    """
    Yığın Sıralaması (Heap Sort) algoritmasını çalıştırır.

    Algoritma 2 adımdan oluşur:
    1. Build_Max_Heap: Diziyi bir Maks-Yığın'a dönüştürür.
    2. Döngü:
       a. Yığının en büyük elemanını (A[0]) alıp dizinin sonuna (A[i]) taşır.
       b. Yığının boyutunu 1 azaltır.
       c. Kök (A[0]) için Max_Heapify çağırarak yığın özelliğini onarır.
    """
    # Başlangıçta yığının mantıksal boyutu tüm dizinin boyutudur.
    heap_size = len(A)

    # 1. Adım: Diziyi bir Maks-Yığın'a dönüştür.
    # Bu işlemden sonra A[0] dizideki en büyük eleman olur.
    Build_Max_Heap(A, heap_size)

    # 2. Adım: Yığından elemanları tek tek çıkar ve sırala
    # Döngü, dizinin son elemanından (n-1) başlayarak
    # ikinci elemanına (indeks 1) kadar geriye doğru gider.
    # (indeks 0'a gitmeye gerek yoktur, son eleman zaten yerindedir).
    for i in range(len(A) - 1, 0, -1):
        # a. En büyük elemanı (A[0]) al, dizinin sıralanmamış
        #    bölümünün sonuna (A[i]) taşı.
        temp = A[0]
        A[0] = A[i]
        A[i] = temp

        # b. Yığının mantıksal boyutunu 1 azalt.
        #    (A[i] artık sıralanmış bölgededir ve yığının parçası değildir).
        heap_size = heap_size - 1

        # c. Yer değiştirme sonucu A[0]'a gelen yeni eleman
        #    yığın özelliğini bozdu. Yığını kökten (0) başlayarak
        #    ve azalan boyuta göre (heap_size) onar.
        Max_Heapify(A, heap_size, 0)

    # Not: Bu fonksiyon doldurulduğunda, Max_Heapify özyineli olsa bile
    # derinliği O(log n) olacağı için stack sorunu yaşatmaz.


# --- ANA (main) FONKSİYON ---

# ... [Diğer tüm fonksiyonlar (InsertionSort, MergeSort, Quicksort, Heap_Sort)
# ...  önceki cevaptaki DÜZELTİLMİŞ halleriyle aynı kalacak] ...


def main():
    """C'deki main() fonksiyonunun Python karşılığı."""

    A1 = []
    random.seed(time.time())

    for _ in range(ARRAY_SIZE):
        A1.append(random.randint(0, 2 ** 31 - 1))

    # Diğer dizileri A1'in *kopyası* olarak oluştur
    A2 = A1.copy()
    A3 = A1.copy()
    A4= A1.copy()  # Heap Sort (Sırasız) için

    print(f"Dizi Boyutu: {ARRAY_SIZE}\n")

    # --- Eklemeli Sıralama ---
    #start_t = time.perf_counter()
    #InsertionSort(A1)
    #end_t = time.perf_counter()
    #cpu_time_used = end_t - start_t
    #print(f"Toplam gecen sure (Eklemeli Siralama): {cpu_time_used:.6f} saniyedir.")

    # --- Birleştirmeli Sıralama ---
    start_t = time.perf_counter()
    MergeSort(A2, 0, ARRAY_SIZE - 1)
    end_t = time.perf_counter()
    cpu_time_used = end_t - start_t
    print(f"Toplam gecen sure (Birlestirmeli Siralama): {cpu_time_used:.6f} saniyedir.")

    # --- Hızlı Sıralama ---
    start_t = time.perf_counter()
    QUICKSORT(A3, 0, ARRAY_SIZE - 1)
    end_t = time.perf_counter()
    cpu_time_used = end_t - start_t
    print(f"Toplam gecen sure (Hizli Siralama): {cpu_time_used:.6f} saniyedir.")

    # --- Yığın Sıralama ---
    # C'deki 'for(i = 0; i < ARRAYSIZE; i++)' döngüsü
    #for i in range(ARRAY_SIZE):
        # C'deki 'printf("%d: %d\n", i+1, A4[i]);'
    #    print(f"{i + 1}: {A4[i]}")

    start_t = time.perf_counter()
    Heap_Sort(A4)  # Sıralama işlemi
    end_t = time.perf_counter()
    cpu_time_used = end_t - start_t
    print(f"\nToplam gecen sure (Yigin Siralama): {cpu_time_used:.6f} saniyedir.\n")

    # C'deki 'for(i = 0; i < ARRAYSIZE; i++)' döngüsü
    #for i in range(ARRAY_SIZE):
        # C'deki 'printf("%d: %d\n", i+1, A4[i]);'
    #    print(f"{i + 1}: {A4[i]}")

# Python kodunu çalıştırmak için standart giriş noktası
if __name__ == "__main__":
    # ÖNEMLİ: Diğer tüm sıralama fonksiyonları (Heap_Sort dahil)
    # bu 'main' fonksiyonundan önce tanımlanmış olmalıdır.
    main()
